### PR TITLE
Reduce MpiCallbacks complexity

### PR DESCRIPTION
**Pull request by jngrad**
Follow-up to #4614
Follow-up to https://github.com/espressomd/espresso/issues/3180#issuecomment-1564376850

### CODE REVIEW

File: src/core/observables/utils_histogram.hpp [hunk 1]
```diff
@@ -21,10 +21,36 @@
...
 #include <vector>
 
-namespace Observables {
+namespace Observables::detail {
+
+/** @brief Gather data from all MPI ranks. */
+template <class Pos>
+auto gather(boost::mpi::communicator const &comm,
+            std::vector<Pos> const &local_pos) {
+  std::vector<std::vector<Pos>> global_pos{};
```
Comment (reinaual) [inline src/core/observables/utils_histogram.hpp:38]:
```suggestion
  std::vector<std::vector<Pos>> global_pos{};
  global_pos.reserve(comm.size());
```

File: src/core/observables/utils_histogram.hpp [hunk 2]
```diff
@@ -21,10 +21,36 @@
...
+}
+
+/** @brief Gather data from all MPI ranks. */
+template <class Pos, class Val>
+auto gather(boost::mpi::communicator const &comm,
+            std::vector<Pos> const &local_pos,
+            std::vector<Val> const &local_val) {
+  std::vector<std::vector<Pos>> global_pos{};
+  boost::mpi::gather(comm, local_pos, global_pos, 0);
+  std::vector<std::vector<Val>> global_val{};
```
Comment (reinaual) [inline src/core/observables/utils_histogram.hpp]:
```suggestion
  auto const world_size = comm.size();
  std::vector<std::vector<Pos>> global_pos{};
  global_pos.reserve(world_size);
  boost::mpi::gather(comm, local_pos, global_pos, 0);
  std::vector<std::vector<Val>> global_val{};
  global_val.reserve(world_size);
```

File: src/core/observables/utils_histogram.hpp (hunk 1 above)
Comment (reinaual) [reply src/core/observables/utils_histogram.hpp:38]: Is there a reason to not take over the `reserve`-s you added in the last PR?

Comment (jngrad) [reply src/core/observables/utils_histogram.hpp:38]: my bad. Getting gcc-10 to properly capture line coverage information for `std::vector` constructors took some trial and error. Let's add those `.reserve()` back.

**Review by reinaual — APPROVED** [review]

**Pull request merged**